### PR TITLE
[PS-2185] Migrate to Microsoft.Data.SqlClient

### PR DESCRIPTION
**Pull request by withinfocus**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Migrate to the new(-ish) SQL client maintained by Microsoft per 
https://github.com/dotnet/SqlClient/blob/main/porting-cheat-sheet.md.

## Code changes

Various C# classes utilizing the new namespace, but really only 
**src/Infrastructure.Dapper/Infrastructure.Dapper.csproj** with a move from `System.Data.SqlClient` to the latest 
`Microsoft.Data.SqlClient`.

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

**Comment by kspearrin** [comment]
Some issues with your package lock files. Check CI failure

**Comment by justindbaur** [comment]
Good timing, I made a [ticket](https://bitwarden.atlassian.net/browse/PS-2185) for this the other day. One thing that will need to be tested thoroughly is that `Microsoft.Data.SqlClient` has a default value of `Encrypt=true` which also requires that the SQL Server needs to have a trusted certificate OR `TrustServerCertificate=True` has to be set in the connection string. We'll want to make sure this won't have any service disruptions for self hosted installations.

**Review by justindbaur — COMMENTED** [review]
Should we remove the reference to the old package? 
https://github.com/bitwarden/server/blob/e7958609b68c996c308695e6132bc1457fed9e57/src/Infrastructure.Dapper/Infrastructure.Dapper.csproj#L9

**Comment by withinfocus** [comment]
Isolated testing is needed for this one, so I wouldn't want to merge it until we have a plan in place.

File: src/Infrastructure.Dapper/Infrastructure.Dapper.csproj [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
     <ItemGroup>
       <PackageReference Include="Dapper" Version="2.0.123" />
-      <PackageReference Include="System.Data.SqlClient" Version="4.8.5" />
+      <PackageReference Include="Microsoft.Data.SqlClient" Version="5.0.1" />
```
Comment (justindbaur) [inline src/Infrastructure.Dapper/Infrastructure.Dapper.csproj:9]: We should be able to get rid of this totally and rely on the peer dependency in `Core.csproj`

**Review by justindbaur — CHANGES_REQUESTED** [review]

File: src/Infrastructure.Dapper/Infrastructure.Dapper.csproj (hunk 1 above)
Comment (withinfocus) [reply src/Infrastructure.Dapper/Infrastructure.Dapper.csproj:9]: For some reason I thought I wanted to keep this here for clarity of dependencies, but you're right that for efficiency's sake and the clear `Core` dependency that I should get rid of it. Good catch.

**Review by justindbaur — APPROVED** [review]

**Comment by kspearrin** [comment]
> Good timing, I made a [ticket](https://bitwarden.atlassian.net/browse/PS-2185) for this the other day. One thing that will need to be tested thoroughly is that `Microsoft.Data.SqlClient` has a default value of `Encrypt=true` which also requires that the SQL Server needs to have a trusted certificate OR `TrustServerCertificate=True` has to be set in the connection string. We'll want to make sure this won't have any service disruptions for self hosted installations.

Looks like we set this automatically for Setup-installed instances. https://github.com/bitwarden/server/blob/master/util/Setup/EnvironmentFileBuilder.cs#L62-L70

Still, we should probably list it in the release notes.

**Comment by withinfocus** [comment]
Logged a release notes addition in DHCTW-866.

**Pull request merged**